composer/models/loss.py

```diff
@@ -199,20 +228,31 @@ def soft_cross_entropy(input: Tensor,
         assert reduction in ['sum', 'mean', 'none'], f"{reduction} reduction not supported."
         assert size_average is None, "size_average is deprecated"
         assert reduce is None, "reduce is deprecated"
-        assert ignore_index == -100, "ignore_index not supported."
-        probs = -1 * (target * F.log_softmax(input, dim=1))
+        if ignore_index != -100:
+            warnings.warn("ignore_index not supported when using dense labels. Ignoring targets with 0 probability.")
+        xentropy = -1 * (target * F.log_softmax(input, dim=1))
 
         if weight is not None:
-            probs *= weight / weight.sum()  # allow broadcast along batch dim
+            # Ugly dimension shuffle to make multiplication work.
+            xentropy = torch.movedim(xentropy, 1, -1)
+            xentropy *= weight  # PyTorch doesn't normalize weights
+            xentropy = torch.movedim(xentropy, -1, 1)
 
-        probs = probs.sum(dim=1)
+        xentropy = xentropy.sum(dim=1)
+        num_examples = torch.numel(xentropy)
 
         if reduction == 'sum':
-            probs = probs.sum(dim=0)
+            xentropy = xentropy.sum()
         elif reduction == 'mean':
-            probs = probs.mean(dim=0)
-
-        return probs
+            xentropy = xentropy.mean()
+            # Reweight loss to account for examples with less than 1 total probability (ignored examples)
+            total_prob = target.sum()
+            assert total_prob > 0, "No targets have nonzero probability"
```

Comment:
This isn't a precondition of the function, so we should probably raise a RunTimeError or ValueError here instead of asserting